pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +75,129 @@ type RevisionRollout struct {
+		return cur
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// TODO: optimize in the later passes.
+
+	// Map the configs by tag.
+	currConfigs, prevConfigs := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i, cfg := range cur.Configurations {
+		currConfigs[cfg.Tag] = append(currConfigs[cfg.Tag], &cur.Configurations[i])
+	}
+	for i, cfg := range prev.Configurations {
+		prevConfigs[cfg.Tag] = append(prevConfigs[cfg.Tag], &prev.Configurations[i])
+	}
+
+	var ret []ConfigurationRollout
+	for t, ccfgs := range currConfigs {
+		pcfgs, ok := prevConfigs[t]
+		// A new tag was added, so we have no previous state to roll from,
+		// thus just add it to the return, we'll rollout to 100% from the get go (and it is
+		// always 100%, since default tag is _always_ there).
+		// So just append to the return list.
+		if !ok {
+			ret = append(ret, *ccfgs[0])
+			continue
+		}
+		// This is basically an intersect algorithm,
+		// It relies on the fact that inputs are sorted.
+		i := 0
+		for j := 0; i < len(ccfgs) && j < len(pcfgs); {
+			switch {
+			case ccfgs[i].ConfigurationName == pcfgs[j].ConfigurationName:
+				// Config might have 0% traffic assigned, if it is a tag only route (i.e.
+				// receives no traffic via default tag).
+				if ccfgs[i].Percent != 0 {
+					ret = append(ret, *stepConfig(ccfgs[i], pcfgs[j]))
+				}
+				i++
+				j++
+			case ccfgs[i].ConfigurationName < pcfgs[j].ConfigurationName:
+				// A new config, has been added. No action for rollout though.
+				ret = append(ret, *ccfgs[i])
+				i++
+			default: // cur > prev.
+				// A config has been removed during this update.
+				// Again, no action for rollout, since this will no longer
+				// be rolling it out (or sending traffic to it overall).
+				j++
+			}
+		}
+		// Keep the remaining new objects
+		for ; i < len(ccfgs); i++ {
+			ret = append(ret, *ccfgs[i])
+		}
+	}
+	ro := &Rollout{Configurations: ret}
+	// We need to sort the rollout, since we have map iterations in between,
+	// which are random.
+	sortRollout(ro)
+	return ro
+}
+
+// stepConfig takes previous and goal configuration shapes and returns a new
+// config rollout, after computing the percetage allocations.
+func stepConfig(goal, prev *ConfigurationRollout) *ConfigurationRollout {
+	pc := len(prev.Revisions)
+	ret := *goal
+	// goal will always have just 1 element – the current desired revision.
+	if goal.Revisions[0].RevisionName == prev.Revisions[pc-1].RevisionName {
+		// TODO(vagababov): here would go the logic to compute new percentages for the rollout,
+		// i.e step function, so return value will change, depending on that.
+		// TODO(vagababov): percentage might change, so this should trigger recompute of existing
+		// revision rollouts.
+		return &ret
+	}
+
+	// Append the new revision, to the list of previous ones.
+	// This is how we start the rollout.
+	rev := goal.Revisions[0]
```

Comment:
I feel like hoisting this up above and calling it `goalRev` is what the above comment is really meant for. 🤔

Reply:
`// goal will always have just 1 element – the current desired revision`

Reply:
🤷 
why move it up? I moved it down.